### PR TITLE
Add support for right-click in GLFW.

### PR DESCRIPTION
**Pull request by ryanknauer**
GLFW currently does not support right-click events for Flutter. This adds buttons to FlutterPointerEvents sent from GLFW.

flutter/flutter#77943

#25074

This is attempting to copy the functionality from MacOS here:

https://github.com/flutter/engine/blob/master/shell/platform/darwin/macos/framework/Source/FlutterViewController.mm#L424

### CODE REVIEW

**Review by rokob — COMMENTED** [review]
LGTM, but I added Greg to double check.

File: shell/platform/glfw/flutter_glfw.cc
Comment (stuartmorgan) [inline shell/platform/glfw/flutter_glfw.cc:68]: Please comment these fields, as with the other fields.

File: shell/platform/glfw/flutter_glfw.cc
Comment (stuartmorgan) [inline shell/platform/glfw/flutter_glfw.cc:342]: Please change the comment on this method, since it's no longer just the primary button. It would also be good to comment that `buttons` needs to be set *before* calling this method. In fact, maybe make `buttons` an argument to make that much more strongly enforced.

File: shell/platform/glfw/flutter_glfw.cc
Comment (stuartmorgan) [inline shell/platform/glfw/flutter_glfw.cc:377]: Why was this moved? You're not using it in the new code.

**Review by stuartmorgan — CHANGES_REQUESTED** [review]
(As a general note on testing, the GLFW code was "temporary" code that has far, far outlived its expected lifespan. If any more non-trivial logic needs to be added, it would be good to do some refactoring for testability first.)

File: shell/platform/glfw/flutter_glfw.cc
Comment (ryanknauer) [reply shell/platform/glfw/flutter_glfw.cc]: Updated.

File: shell/platform/glfw/flutter_glfw.cc
Comment (ryanknauer) [reply shell/platform/glfw/flutter_glfw.cc:346]: Updated.

File: shell/platform/glfw/flutter_glfw.cc
Comment (ryanknauer) [reply shell/platform/glfw/flutter_glfw.cc:69]: Updated.

File: shell/platform/glfw/flutter_glfw.cc
Comment (chunhtai) [inline shell/platform/glfw/flutter_glfw.cc:352]: I am confused at the logic here.
if buttons evaluate to true (at lease on button is down) and controller->pointer_currently_down, shouldn't we issue move? It feel to me the condition should be reverted.

File: shell/platform/glfw/flutter_glfw.cc
Comment (chunhtai) [inline shell/platform/glfw/flutter_glfw.cc:328]: how do we keep track of which of the two button is down if we only use 1 boolean?

For example, if left-click down and right click down and you lifted left click, the controller->pointer_currently_down will become false

Comment (ryanknauer) [reply shell/platform/glfw/flutter_glfw.cc:328]: I messed up the conditional here a bit in a clean up commit, it is supposed to be (buttons == 0). I think your question still applies though:

Basically this is really ugly, but the engine doesn't differentiate between left and right mouse buttons when setting the kDown and kUp states. The buttons field is basically just additional field for telling which buttons are currently down. This is basically a copy of what is happening in MacOS here:

https://github.com/flutter/engine/blob/master/shell/platform/darwin/macos/framework/Source/FlutterViewController.mm#L424

The kUp and kDown events here are only called if the stored buttons state moves from no buttons presses (buttons == 0) to any buttons pressed, or vice versa. 

So, in the case you describe - when the left click is lifted the SetEventPhaseFromCursorButtonState will trigger a kMove phase with the buttons being updated to just the right mouse button. This is obviously not entirely correct, but it seems like the best option and is a mirror of MacOS.

Comment (chunhtai) [reply shell/platform/glfw/flutter_glfw.cc:328]: Oh yes, I forgot about this, lifting one finger while pressing down another should issue a move, and as long as that is the case in linux embedding that should be fine. I think what you have right now is correct.

**Review by chunhtai — COMMENTED** [review]
logic looks good, but we need a test in order to merge this pr.

**Comment by chinmaygarde** [comment]
It seems like both @stuartmorgan s and @chunhtai s comments have been addressed and the functionality looks correct. Is the [insistence on the test](https://github.com/flutter/engine/pull/25110#pullrequestreview-616847618) related to updating an existing test, adding a new test to an existing harness, or inventing a test harness to add a new test. If its the final condition, maybe we need to track that separately and unblock issues of this ilk.

**Comment by stuartmorgan** [comment]
Given that the changes to the GLFW embedding tend to be infrequent and small, and backfilling test infrastructure for it has been *very* low priority, I'm not sure requiring a test here is actually necessary unless there are very likely to be more follow-up PRs adding functionality. (There may not be a way to inject test mouse events without refactoring to make the handlers accessible outside the file, which is doable, but refactoring for testability would likely be significantly larger than the PR).

That's ultimately up to @cbracken to decide though.

**Comment by cbracken** [comment]
I'm fine with passing on the test in this case. As @stuartmorgan notes, backfilling a harness is going to be low priority.

**Review by cbracken — APPROVED** [review]

**Pull request merged**
